Doc: Update for the "extras" section in pyproject.toml file (#3156)
python-poetry/poetry: docs/docs/pyproject.md

# below `extras`. They can be opted into by apps.
psycopg2 = { version = "^2.7", optional = true }
mysqlclient = { version = "^1.3", optional = true }

> rich = { version = "^8.0", optional = true }
> emoji = { version = "^0.6", optional = true }

Reviewer: ```suggestion
```

[tool.poetry.extras]
mysql = ["mysqlclient"]
pgsql = ["psycopg2"]

> styles = ["rich", "emoji"]

Reviewer: ```suggestion
databases = ["mysqlclient", "psycopg2"]
```

```

`extras` can only be used with packages defined in the `dependencies` section.

> When pip installing the package, the dependencies specified in the `extras` section can be installed by providing

Reviewer: We should really be talking about `pip` as an example of a PEP 517 front-end, not necesarily the defacto option that uses this. The conversation should also probably reference PEP 508 when mentioning how extras are consumed. Especially because "extras" is not something specific to poetry.

Reviewer 2: Great point! I will update this

[tool.poetry.extras]
mysql = ["mysqlclient"]
pgsql = ["psycopg2"]
databases = ["mysqlclient", "psycopg2"]
```

> 

Reviewer: Maybe move this section after the `poetry` specific documentation.

Author: Just so I am clear, do you mean move this to [section](https://python-poetry.org/docs/pyproject/#poetry-and-pep-517)?

Reviewer: I should have been clearer, sorry. I was proposing that your additions go blow what already exists as opposed to above it. So your new lines regarding `x[a]` usage go after the existing documentation around using `poetry install -E`.

```bash
poetry install --extras "mysql pgsql"
poetry install -E mysql -E pgsql
```

> When installing or specifying poetry built packages, the extras defined in this section can be activated 

Reviewer 3: ```suggestion
When installing or specifying poetry-built packages, the extras defined in this section can be activated 
```

> When using build front-ends that comply with [PEP-517](https://www.python.org/dev/peps/pep-0517/), the dependencies 
> specified in the `extras` section, can be installed by providing the key between `[]` along with the package name 
> (as prescribed in [PEP-508](https://www.python.org/dev/peps/pep-0508/#extras)). 
> 
> For example, from the above toml file, to install the extras under the `databases` key using `pip` as the build front-end

Reviewer: ```suggestion
When installing or specifying poetry built packages, the extras defined in this section can be activated as described in [PEP 508](https://www.python.org/dev/peps/pep-0508/#extras).

For example, the dependencies requiredf by the `databases` extra can be activated when installing the package using `pip` as shown below.
```

I think we can be a little less verbose here.

as described in [PEP 508](https://www.python.org/dev/peps/pep-0508/#extras).

> For example, the dependencies required by the `databases` extra can be activated when installing the 

Reviewer 4: Is "activated" the right word here? Doesn't "installed" more accurately describe what actually happens?

Author: Good catch! Will get it fixed

package using `pip` as shown below.

```bash
pip install awesome[databases]
```

!!!note

>     The dependencies specified for each extra must be already defined as project dependencies.

Reviewer 3: ```suggestion
    The dependencies specified for each `extra` must be already defined as project dependencies.
```

>     The dependencies specified for each `extra` must be already defined as project dependencies.

Reviewer 4: "be already" → "already be" (sounds more natural in this order)

>     Using any dependency from the dev dependencies section is not supported.

Reviewer 3: ```suggestion
    Using any dependency from the `dev-dependencies` section is not supported.
```

>     Using any dependency from the `dev-dependencies` section is not supported.

Reviewer 4: This sentence sounds a bit awkward. Perhaps the following phrasing instead?
> Dependencies listed in the `dev-dependencies` section cannot be specified as extras.

> 

Author: @abn Wouldn't it be better to explicitly state this or is this note not true anymore?

> 

Reviewer: Might be good to rephrase it I reckon. Right now its a bit a unclear, hence why I proposed to remove it. If you word it along the lines of "the dependencies specified for each extra must already be defined as project dependencies".

> 

Author: got it 👍🏽

databases = ["mysqlclient", "psycopg2"]
```

When installing packages with poetry, you can specify extras by using the `-E|--extras` option:
When installing packages with Poetry, you can specify extras by using the `-E|--extras` option:

```bash
poetry install --extras "mysql pgsql"
poetry install -E mysql -E pgsql
```

When installing or specifying poetry-built packages, the extras defined in this section can be activated 
When installing or specifying Poetry-built packages, the extras defined in this section can be activated 
as described in [PEP 508](https://www.python.org/dev/peps/pep-0508/#extras).

For example, the dependencies required by the `databases` extra can be activated when installing the 
package using `pip` as shown below.
For example, when installing the package using `pip`, the dependencies required by 
the `databases` extra can be installed as shown below.

```bash
pip install awesome[databases]
```

!!!note

>     The dependencies specified for each `extra` must be already defined as project dependencies.

Reviewer 4: Did you mean to remove this sentence? I think it's better to leave it, but with the "be already" → "already be" swapped.

Author: Oops that slipped past me, haven't yet had my cup of tea for the day 😁
